go/pkg/services/logcollector/server.go

```diff
@@ -0,0 +1,605 @@
+			if logFilePath == "" || info.ModTime().After(latestModTime) {
+				logFilePath = path
+				latestModTime = info.ModTime()
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return "", errors.Wrap(err, "Failed to list files in base directory")
+	}
+
+	// if no log file was found, return error
+	if logFilePath == "" {
+		return "", errors.Errorf("Failed to find log file for run id %s", runUID)
+	}
+
+	return logFilePath, nil
+}
+
+func (s *LogCollectorServer) readLogsFromFile(ctx context.Context,
+	runUID,
+	filePath string,
+	offset uint64,
+	size int64) ([]byte, error) {
+
+	// open log file for reading
+	file, err := os.OpenFile(filePath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to open log file for run id %s", runUID)
+	}
+	defer file.Close() // nolint: errcheck
+
+	// get file size
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get file info for run id %s", runUID)
+	}
+
+	offset, size = s.validateOffsetAndSize(offset, size, fileInfo.Size())
+	if size == 0 {
+		s.Logger.DebugWithCtx(ctx, "No logs to return", "run_id", runUID)
+		return nil, nil
+	}
+
+	// read size bytes from offset
+	buffer := make([]byte, size)
+	if _, err := file.ReadAt(buffer, int64(offset)); err != nil {
+
+		// if error is EOF, return empty bytes
+		if err == io.EOF {
+			return buffer, nil
+		}
+
+		// else return error
+		err := errors.Wrapf(err, "Failed to read log file for run id %s", runUID)
+		return nil, err
+	}
+
+	return buffer, nil
+}
+
+func (s *LogCollectorServer) validateOffsetAndSize(offset uint64, size, fileSize int64) (uint64, int64) {
+
+	// if size is negative, zero, or bigger than fileSize, read the whole file
+	if size <= 0 || size > fileSize {
+		size = fileSize
+	}
+
+	// if offset is bigger than file size, set offset to 0
+	if int64(offset) > fileSize {
+		offset = 0
+	}
+
+	// set size to file size - offset if size is bigger than file size - offset
+	if size > fileSize-int64(offset) {
+		size = fileSize - int64(offset)
+	}
+
+	return offset, size
+}
+
+func (s *LogCollectorServer) monitorLogCollection(ctx context.Context) {
+
+	s.Logger.DebugWithCtx(ctx,
+		"Monitoring log streaming goroutines periodically",
+		"monitoringInterval", s.monitoringInterval)
+
+	monitoringTicker := time.NewTicker(s.monitoringInterval)
+
+	// Check the items in the inMemoryState against the items in the state store.
+	// If an item is written in the state store but not in the in memory state - call StartLog for it,
+	// as the state store is the source of truth
+	for range monitoringTicker.C {
+
+		// if there are already log items in progress, call StartLog for each of them
+		logItemsInProgress, err := s.stateStore.GetItemsInProgress()
+		if err == nil {
+			logItemsInProgress.Range(func(key, value any) bool {
+				runUID, ok := key.(string)
+				if !ok {
+					s.Logger.WarnWithCtx(ctx, "Failed to convert runUID key to string")
```

Comment:
i am not sure i follow. why converting to str? and if you fail? doesnt look like you're doing something here

Reply:
> why converting to str?
When using `sync.Map` the key and values are type-less. In order to use them for the following `Load` and `StartLog`, they must be converted to their real type.

> and if you fail? doesnt look like you're doing something here
You're right, that's by mistake. I will continue iterating and not fail, since ther might be other items to start log for, and we don't really want to fail the entire server.